### PR TITLE
device: add Window.set_cursor_mode()

### PR DESCRIPTION
**Pull request by vasconssa**

### CODE REVIEW

File: samples/01-physics/main.lua [hunk 1]
```diff
@@ -30,6 +30,7 @@ function Game.level_loaded()
 	-- Debug
 	PhysicsWorld.enable_debug_drawing(Game.pw, debug_physics)
 	RenderWorld.enable_debug_drawing(Game.rw, debug_graphics)
+	Window.set_cursor_mode("disabled")
```
Comment (dbartolini) [inline samples/01-physics/main.lua]: Remove this. Start in `normal` mode.

File: src/device/main_android.cpp [hunk 2]
```diff
@@ -271,6 +271,9 @@ struct WindowAndroid : public Window
 	void set_cursor(MouseCursor::Enum /*cursor*/)
 	{
 	}
+	virtual void set_cursor_mode(CursorMode::Enum mode)
```
Comment (dbartolini) [inline src/device/main_android.cpp]: Remove `virtual`. Comment-out `mode` to silence warning.

File: src/device/main_linux.cpp [hunk 3]
```diff
@@ -472,13 +493,36 @@ struct LinuxDevice
...
+									, MouseAxis::CURSOR
+									, window_context.virtual_cursor_posx
+									, window_context.virtual_cursor_posy
+									, 0
+									);
+								XWarpPointer(_x11_display, None, window_context._x11_window,
+										0, 0, 0, 0, window_context.width/2, window_context.height/2);
+								window_context.last_cursor_posx = window_context.width/2;
+								window_context.last_cursor_posy = window_context.height/2;
+								printf("lastx: %d, lasty: %d\n", window_context.last_cursor_posx, window_context.last_cursor_posy);
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Remove printf().

File: src/device/main_linux.cpp [hunk 4]
```diff
@@ -472,13 +493,36 @@ struct LinuxDevice
...
+								window_context.virtual_cursor_posy += y - window_context.last_cursor_posy;
+								_queue.push_axis_event(InputDeviceType::MOUSE
+									, 0
+									, MouseAxis::CURSOR
+									, window_context.virtual_cursor_posx
+									, window_context.virtual_cursor_posy
+									, 0
+									);
+								XWarpPointer(_x11_display, None, window_context._x11_window,
+										0, 0, 0, 0, window_context.width/2, window_context.height/2);
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Add XFlush() right after XWarpPointer(). Fix indentation.

File: src/device/main_linux.cpp [hunk 5]
```diff
@@ -472,13 +493,36 @@ struct LinuxDevice
...
-						, 0
-						, MouseAxis::CURSOR
-						, event.xmotion.x
-						, event.xmotion.y
-						, 0
-						);
+					{
+						const s32 x = event.xmotion.x;
+						const s32 y = event.xmotion.y;
+						if (window_context.cursor_mode == CursorMode::DISABLED) {
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Fix braces.

File: src/device/main_linux.cpp [hunk 6]
```diff
@@ -523,6 +567,7 @@ struct LinuxDevice
 				default:
 					break;
 				}
+				XFlush(_x11_display);
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Do not call XFlush() with a light heart, call it only when necessary (i.e. only right after specific X11 calls like above).

File: src/device/main_linux.cpp [hunk 7]
```diff
@@ -590,6 +633,11 @@ struct WindowX11 : public Window
 		::Window root_window = RootWindow(s_ldvc._x11_display, screen);
 		::Window parent_window = (parent == 0) ? root_window : (::Window)parent;
 
+		window_context.width = width;
+		window_context.height = height;
+		window_context.recover_cursor_posx = width/2;
+		window_context.recover_cursor_posy = height/2;
+
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Remove this. See below.

File: src/device/main_linux.cpp [hunk 8]
```diff
@@ -629,23 +677,24 @@ struct WindowX11 : public Window
...
 			);
-		CE_ASSERT(_x11_window != None, "XCreateWindow: error");
+		CE_ASSERT(window_context._x11_window != None, "XCreateWindow: error");
 
-		XSetWMProtocols(s_ldvc._x11_display, _x11_window, &s_ldvc._wm_delete_window, 1);
+		XSetWMProtocols(s_ldvc._x11_display, window_context._x11_window, &s_ldvc._wm_delete_window, 1);
 
-		XMapRaised(s_ldvc._x11_display, _x11_window);
+		XMapRaised(s_ldvc._x11_display, window_context._x11_window);
+		set_cursor_mode((CursorMode::Enum)window_context.cursor_mode);
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Remove this. Cursor mode is normal by default anyway.

File: src/device/main_linux.cpp [hunk 9]
```diff
@@ -743,7 +792,50 @@ struct WindowX11 : public Window
...
+
+		if (mode == CursorMode::DISABLED) {
+			::Window root_return, child_return;
+			int root_x_return, root_y_return, win_x_return, win_y_return;
+			unsigned int mask;
+			XQueryPointer(s_ldvc._x11_display, window_context._x11_window, &root_return,
+					&child_return, &root_x_return, &root_y_return, &win_x_return, &win_y_return,
+					&mask);
+			window_context.recover_cursor_posx = win_x_return;
+			window_context.recover_cursor_posy = win_y_return;
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Do not restore original mouse coords. Does it actually make sense?

File: src/device/main_linux.cpp [hunk 10]
```diff
@@ -743,7 +792,50 @@ struct WindowX11 : public Window
...
+
+			XWarpPointer(s_ldvc._x11_display, None, window_context._x11_window,
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Do not use cached width/height here, call XGetWindowAttributes() to retrieve window dimensions.

File: src/device/main_linux.cpp [hunk 11]
```diff
@@ -743,7 +792,50 @@ struct WindowX11 : public Window
...
+			window_context.virtual_cursor_posy = window_context.height/2;
+
+			// needed case engine begins in NORMAL mode
+			s_ldvc._queue.push_axis_event(InputDeviceType::MOUSE
+				, 0
+				, MouseAxis::CURSOR
+				, window_context.width/2
+				, window_context.height/2
+				, 0
+				);
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Remove this. _queue is SPSC, all push() calls must occur on the same thread (i.e. the main thread).
Window::set_cursor_mode() may be called from any thread.

File: scripts/openal.lua [hunk 12]
```diff
@@ -101,6 +101,8 @@ function openal_project(_kind)
 				"HAVE_WASAPI",
 				"HAVE_WINDOWS_H",
 				"HAVE_WINMM",
+				"strcasecmp=_stricmp",
+				"strncasecmp=_strnicmp",
```
Comment (dbartolini) [inline scripts/openal.lua]: This does not belong here, I will fix it in master.

File: samples/01-physics/main.lua [hunk 13]
```diff
@@ -70,9 +71,9 @@ function Game.update(dt)
 
 	-- Update camera
 	local delta = Vector3.zero()
-	if Mouse.button(Mouse.button_id("right")) > 0 then
-		delta = Mouse.axis(Mouse.axis_id("cursor_delta"))
-	end
+	--if Mouse.button(Mouse.button_id("right")) > 0 then
+	delta = Mouse.axis(Mouse.axis_id("cursor_delta"))
+	--end
```
Comment (dbartolini) [inline samples/01-physics/main.lua]: Toggle between `normal` and `disabled` with keyboard spacebar.

File: src/device/main_linux.cpp [hunk 14]
```diff
@@ -297,6 +297,21 @@ struct Joypad
...
+	s16 virtual_cursor_posy = 0;
+	s16 last_cursor_posx = 0;
+	s16 last_cursor_posy = 0;
+	s16 recover_cursor_posx = 0;
+	s16 recover_cursor_posy = 0;
+	u8 cursor_mode = CursorMode::NORMAL;
+};
+static WindowContext window_context;
+
+
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: There is no need for WindowContex. Move its members into LinuxDevice (X11Window can access it) and initialize with LinuxDevice ctor.

File: src/device/main_linux.cpp [hunk 15]
```diff
@@ -416,11 +431,15 @@ struct LinuxDevice
 				switch (event.type)
 				{
 				case EnterNotify:
+					window_context.last_cursor_posx = window_context.width/2;
+					window_context.last_cursor_posy = window_context.height/2;
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: I'm not sure why we need this?

File: src/device/main_linux.cpp [hunk 16]
```diff
@@ -416,11 +431,15 @@ struct LinuxDevice
...
 					_queue.push_axis_event(InputDeviceType::MOUSE
 						, 0
 						, MouseAxis::CURSOR
-						, event.xcrossing.x
-						, event.xcrossing.y
+						, window_context.cursor_mode == CursorMode::DISABLED ?
+														window_context.width/2 : event.xcrossing.x
+						, window_context.cursor_mode == CursorMode::DISABLED ?
+														window_context.height/2 : event.xcrossing.y
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Ditto.

File: src/device/main_linux.cpp [hunk 17]
```diff
@@ -431,6 +450,8 @@ struct LinuxDevice
 					break;
 
 				case ConfigureNotify:
+					window_context.width = event.xconfigure.width;
+					window_context.height = event.xconfigure.height;
```
Comment (dbartolini) [inline src/device/main_linux.cpp]: Is there a way to avoid caching width/height and achieve same effect?

File: src/device/main_android.cpp (hunk 2 above)
Comment (vasconssa) [reply src/device/main_android.cpp]: Ok

File: samples/01-physics/main.lua (hunk 13 above)
Comment (vasconssa) [reply samples/01-physics/main.lua]: i think is better too

**Pull request merged**